src/shared.c

```diff
@@ -696,248 +707,557 @@ bool write_edited_desktop_file(GKeyFile *key_file_structure, const char* appimag
-                fprintf(stderr, "%s", archive_error_string(a));
+                fprintf(stderr, "%s", archive_error_string(*a));
                 break;
             }
-            GKeyFile *key_file_structure = g_key_file_new(); // A structure that will hold the information from the desktop file
-            gboolean success = g_key_file_load_from_data(key_file_structure, buff, size, G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS, NULL);
-            if(success){
-                gchar *desktop_filename = g_path_get_basename(filename);
-                desktop_icon_value_original = g_key_file_get_string(key_file_structure, "Desktop Entry", "Icon", NULL);
-                if(verbose)
-                    fprintf(stderr, "desktop_icon_value_original: %s\n", desktop_icon_value_original);
-                if (!write_edited_desktop_file(key_file_structure, path, desktop_filename, 1, md5, verbose)) {
-                    fprintf(stderr, "Failed to install desktop file\n");
-                    return false;
-                }
 
-                g_free(desktop_filename);
+            *desktop_filename = g_strdup(g_path_get_basename(filename));
+
+            // a structure that will hold the information from the desktop file
+            *key_file = g_key_file_new();
+
+            gboolean success = g_key_file_load_from_data(*key_file, buff, size,
+                G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS, NULL);
+
+            if (!success) {
+                // cleanup
+                g_free(key_file);
+                key_file = NULL;
+
+                break;
             }
-            g_key_file_free(key_file_structure);
+
+            // cleanup
+            g_free(filename);
+
+            return true;
         }
+    }
 
-        gchar *dest = NULL;
-        /* Get icon file(s) and act on them in one go */
+    g_free(filename);
 
-        if(g_str_has_prefix(filename, "usr/share/icons/") || g_str_has_prefix(filename, "usr/share/pixmaps/") || (g_str_has_prefix(filename, "usr/share/mime/") && g_str_has_suffix(filename, ".xml"))){
-            gchar *t = replace_str(filename, "usr/share", g_get_user_data_dir());
-            gchar *dest_dirname = g_path_get_dirname(t);
-            g_free(t);
+    return false;
+}
 
-            gchar *file_basename = g_path_get_basename(filename);
-            gchar *dest_basename = g_strdup_printf("%s_%s_%s", vendorprefix, md5, file_basename);
+bool archive_copy_icons_recursively_to_destination(struct archive** a, const gchar* md5,
+                                                   const gchar* desktop_icon_value_original, gboolean verbose) {
+    // iterate over all files ("entries") in the archive
+    // looking for a file with .desktop extension in the root directory
+    struct archive_entry* entry;
+    gchar* filename = NULL;
 
-            dest = g_build_path("/", dest_dirname, dest_basename, NULL);
+    bool errored = false;
+
+    int r = -1;
+
+    for (;;) {
```

Comment:
Using a while loop would net be more expressive here ?

Reply:
This is pretty much taken unmodified from the old code.